chore: add defensive check to ensure metadata does not change when reopening an active channel

colin-axner opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #795 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

colin-axner commented on modules/apps/27-interchain-accounts/controller/keeper/handshake.go:
```diff
@@ -49,9 +50,20 @@ func (k Keeper) OnChanOpenInit(
+			return sdkerrors.Wrapf(icatypes.ErrActiveChannelAlreadySet, "existing active channel %s for portID %s", activeChannelID, portID)
+		}
+
+		if channel.Version != version {
+			return sdkerrors.Wrap(icatypes.ErrInvalidVersion, "previous active channel metadata does not match provided version")
+		}
```
this requires the full metadata to be passed in when reopening a channel

damiannolan commented on modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:
```diff
@@ -30,6 +30,24 @@ func (suite *KeeperTestSuite) TestOnChanOpenInit() {
+					Ordering:       channeltypes.ORDERED,
+					Counterparty:   counterparty,
+					ConnectionHops: []string{path.EndpointA.ConnectionID},
+					Version:        TestVersion,
```
Hmm, shouldn't the version set in state on the closed channel also include the populated address field? 

This would mean that you cannot re-open a channel using `RegisterInterchainAccount` tho as the metadata built explicitly with an address: https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/account.go#L45-L53

damiannolan replied:
(whoops, didn't see marked as draft)

colin-axner replied:
I had opened it as r4r and then decided to change it up

colin-axner commented:
I reworked the code, checks now added for controller and host. The check added cannot be reached in tests since we don't support multiple application versions.

Running into an issue with the reopening a channel after closure test. Will debug monday, feel free to push to this branch if you see the fix

crodriguezvega commented on modules/apps/27-interchain-accounts/controller/keeper/handshake.go:
```diff
@@ -49,9 +50,20 @@ func (k Keeper) OnChanOpenInit(
+			return sdkerrors.Wrapf(icatypes.ErrActiveChannelAlreadySet, "existing active channel %s for portID %s", activeChannelID, portID)
+		}
+
+		if !icatypes.IsPreviousMetadataEqual(version, metadata) {
```
I might be missing something, but the `version` that is passed here, isn't it the same that is unmarshalled to `metadata` above? So here:

```
var metadata icatypes.Metadata
if err := icatypes.ModuleCdc.UnmarshalJSON([]byte(version), &metadata); err != nil {
    return sdkerrors.Wrapf(icatypes.ErrUnknownDataType, "cannot unmarshal ICS-27 interchain accounts metadata")
}
```

If that's the case, wouldn't this check here just redundant?

Oh, wait... Or shouldn't you be using here `channel.Version` instead?

colin-axner replied:
nice catch, it should be channel.Version :+1:

colin-axner replied:
fixed

crodriguezvega commented on modules/apps/27-interchain-accounts/types/metadata.go:
```diff
@@ -27,6 +27,21 @@ func NewMetadata(version, controllerConnectionID, hostConnectionID, accAddress,
 
+// IsMetadataEqual compares a metadata to a pevious version string set in a channel struct.
+// It ensure all fields are equal except the Address string
+func IsPreviousMetadataEqual(previousVersion string, metadata Metadata) bool {
```
Suggestion...

We already have code to unmarhsall the version string [here](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L43-L46) and the same code is found in [`OnChannelOpenTry`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/host/keeper/handshake.go#L41-L44) and [`OnOpenChannelAck`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L76-L79). And now we also have it inside this function. So we could make an extra constructor function and remove the duplication of code in all these places. So basically to create a function with a signature like this: `func NewMetadata(version string) (Metadata, error)`.

And if you do that then you could also change this function so that it has the `Metadata` receiver type and change also the name and signature, so something like this: `func (Metadata m) AreEqual(metadata Metadata) bool`. I leave it up to you if you prefer a pointer receiver type.

The you can invoke these functions like:

```
previousMetadata, err := icatypes.NewMetadata(previousVersion)
if !err {
    return "", err
}
if !metadata.AreEqual(previousMetadata) {
    ...   
}

```

colin-axner replied:
I like your suggestion. `NewMetadata` would need to be named something else to avoid name collision with the traditional usage of `NewMetadata`. `IsEqual` is also a good idea, but we will remove all this logic with the removal of active channels, so I'm leaning towards leaving as is. It makes the equality check explicitly to be used for active channel security and nothing else (otherwise external users may start depending on that API which should be removed in the future)

damiannolan replied:
I prefer @crodriguezvega's suggestion too, but given this will be removed in a couple of months hopefully, then I think we should just leave it as is.

seantking replied:
agree with @damiannolan

crodriguezvega commented on modules/apps/27-interchain-accounts/host/keeper/handshake.go:
```diff
@@ -47,8 +48,20 @@ func (k Keeper) OnChanOpenTry(
 
-	if activeChannelID, found := k.GetOpenActiveChannel(ctx, connectionHops[0], counterparty.PortId); found {
-		return "", sdkerrors.Wrapf(icatypes.ErrActiveChannelAlreadySet, "existing active channel %s for portID %s", activeChannelID, portID)
+	activeChannelID, found := k.GetActiveChannelID(ctx, connectionHops[0], counterparty.PortId)
```
If this whole block of code is the same as the one in `OnChannelOpenInit`, could we extract it into a helper function?

colin-axner replied:
I think it is fine to leave as is/be explicit. There's still a lot of duplicate code between OpenInit and OpenTry

AdityaSripal commented on modules/apps/27-interchain-accounts/host/keeper/handshake.go:
```diff
@@ -47,8 +48,20 @@ func (k Keeper) OnChanOpenTry(
+		}
+
+		if channel.State == channeltypes.OPEN {
+			return "", sdkerrors.Wrapf(icatypes.ErrActiveChannelAlreadySet, "existing active channel %s for portID %s", activeChannelID, portID)
```
```suggestion
			return "", sdkerrors.Wrapf(icatypes.ErrActiveChannelAlreadySet, "existing active channel %s for portID %s is already OPEN", activeChannelID, portID)
```

colin-axner replied:
updated on controller and host

damiannolan commented on modules/apps/27-interchain-accounts/types/metadata.go:
```diff
@@ -27,6 +27,21 @@ func NewMetadata(version, controllerConnectionID, hostConnectionID, accAddress,
 	}
 }
 
+// IsMetadataEqual compares a metadata to a pevious version string set in a channel struct.
+// It ensure all fields are equal except the Address string
```
```suggestion
// IsPreviousMetadataEqual compares a metadata to a previous version string set in a channel struct.
// It ensures all fields are equal except the Address string
```

seantking reviewed (approved):
:handshake:

colin-axner commented:
> Looks great! I'm assuming there isn't error cases because it's not possible for it to even pass at the moment with some other metadata?

correct

colin-axner commented on modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:
```diff
@@ -30,6 +30,24 @@ func (suite *KeeperTestSuite) TestOnChanOpenInit() {
 			true,
 		},
+		{
+			"success - previous active channel closed",
```
Unreachable, but I might as well add one
